### ISSUE TITLE
noam_decay的使用

### ISSUE THREAD

**Issue by imguozhen** [PaddlePaddle/Paddle#18403]
请教个问题，下面代码，noam_decay的加入对不对：
1. 应该写在什么位置：lr_decay = fluid.layers.noam_decay(512, 4000) 是写在下面代码的位置，还是应该写在with fluid.program_guard(train_program, startup_program)里面，是写在test_program = train_program.clone(for_test=True)之前还是之后
2. learning_rate = learning_rate * lr_decay之前需要加什么吗，比如 with fluid.default_main_program()._lr_schedule_guard()
3. optimizer定义好之后，可以重设learning_rate吗，比如 optimizer = fluid.optimizer.Adam(learning_rate=0.2)  optimizer.learning_rate=0.3
4. 我不太清楚的地方在于：实现代码的时候是不是要保证fluid.layers.noam_decay这样的算子只出现在train_program和startup_program中，而不出现在其它test_program中，而在很多代码中我看到optimizer的位置比较早，貌似包含在所有的program，而noam_decay是在optimizer之前，定义学习率的。是不是也可以把noam_decay和optimizer的位置都移到with fluid.program_guard(train_program, startup_program)里的test_program = train_program.clone(for_test=True)之后，optimizer.minimize(loss)之前
—————————————————————————————————————————————
```
learning_rate = 0.2

lr_decay = fluid.layers.noam_decay(512, 4000)
learning_rate = learning_rate * lr_decay

optimizer = fluid.optimizer.Adam(learning_rate=learning_rate)

train_program = fluid.Program()
startup_program = fluid.Program()

with fluid.program_guard(train_program, startup_program):
    img = fluid.layers.data(name='image', shape=[784])
    hidden = fluid.layers.fc(input=img, size=200, act='relu')
    hidden = fluid.layers.dropout(hidden, dropout_prob=0.5)
    loss = fluid.layers.cross_entropy(
		input=fluid.layers.fc(hidden, size=10, act='softmax'),
		label=fluid.layers.data(name='label', shape=[1], dtype='int64'))

    test_program = train_program.clone(for_test=True)
    optimizer.minimize(loss)
```
————————————————————————————————————————————
5. 下面的代码中with fluid.unique_name.guard()是必须的吗，不使用会有什么影响
6. 下面的代码中with fluid.default_main_program()._lr_schedule_guard()是什么意思？在paddle的文档中找不到这个的说明？这是必须的吗？在with fluid.program_guard(train_program, train_startup)里面应该怎么理解？
```
train_program = fluid.Program()
train_startup = fluid.Program()
with fluid.program_guard(train_program, train_startup): 
    with fluid.unique_name.guard(): 
        feed_order, loss, predict, accuracy, num_seqs = create_model(args, num_labels, is_prediction=False)
        lr_decay = fluid.layers.learning_rate_scheduler.noam_decay(256, warmup_steps)
        with fluid.default_main_program()._lr_schedule_guard(): 
            learning_rate = lr_decay * args.learning_rate
        optimizer = fluid.optimizer.Adam(learning_rate=learning_rate)
        optimizer.minimize(loss)
```

**Comment by tink2123** [comment]
1. 写在 with fluid.program_guard(train_program, startup_program)里面，位置没有要求
2. 不需要，具体使用方法可以参考modes库里的样例：https://github.com/PaddlePaddle/models/blob/a3f6ee86dc63ff5d22c08ec226e55f43e1848204/PaddleNLP/dialogue_model_toolkit/dialogue_general_understanding/optimization.py#L53
3. 如果有需要的话是可以修改的，实现方式可参考：
https://github.com/PaddlePaddle/models/blob/a3f6ee86dc63ff5d22c08ec226e55f43e1848204/PaddleNLP/dialogue_model_toolkit/dialogue_general_understanding/optimization.py#L48
4. 是的，可以放在test后面
5. 不是必须的，但是如果使用了py_reader+train_program+test_program 的情况下，需要使用with fluid.unique_name.guard()让train_program 和 test_program 共享参数
6. _lr_schedule_guard 是Paddle内部的一个接口，没用暴露给用户使用。在这段代码里用于实现学习率衰减，如果希望实现该功能，推荐优先使用开放给用户的API，例如 `paddle.fluid.layers.cosine_decay` 等。

**Comment by imguozhen** [comment]
多谢解答~
